lib/pure/cgi.nim

```diff
@@ -96,8 +96,11 @@ iterator decodeData*(allowedMethods: set[RequestMethod] =
   ## data consists of. If the client does not use a method listed in the
   ## `allowedMethods` set, a ``CgiError`` exception is raised.
   let data = getEncodedData(allowedMethods)
-  for key, value in decodeData(data):
-    yield (key, value)
+  try:
+    for (key, value) in uri.decodeQuery(data):
+      yield (key, value)
+  except UriParseError as e:
+    cgiError(e.msg)
```

Comment:
I think this should be `reraise` otherwise you loose the stacktrace context where it occurred; can you please check?

Reply:
I think It has the similar error messages like before.

![image](https://user-images.githubusercontent.com/43030857/98384598-ef751180-2088-11eb-98bc-155413cda42a.png)


Before:
a
1
b
2c
/playground/nim/lib/pure/cgi.nim(115) in
/playground/nim/lib/pure/cgi.nim(67) cgiError
Error: unhandled exception: '&' expected [CgiError]

Reply:
use this:
```nim
raise newException(CgiError, e.msg, e)
```

this attaches `e` to parent of raised exception, so that client code can get the context:
```nim
import std/cgi
import std/strformat

proc exceptionToString(e: ref Exception): string =
  var e = e
  while e!=nil:
    result.add fmt"name: {e.name} msg: {e.msg} stacktrace: {e.getStackTrace}"
    e = e.parent

proc fn1()=
  for ai in decodeData("asdf=bar=baz"):
    echo ai

proc main() =
  try:
    fn1()
  except Exception as e:
    echo e.exceptionToString

main()
```

prints:
```
name: CgiError msg: '&' expected at 8 stacktrace: /Users/timothee/git_clone/nim/timn/tests/nim/all/t11246.nim(26) t11246
/Users/timothee/git_clone/nim/timn/tests/nim/all/t11246.nim(21) main
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/cgi.nim(91) fn1
name: UriParseError msg: '&' expected at 8 stacktrace: /Users/timothee/git_clone/nim/timn/tests/nim/all/t11246.nim(26) t11246
/Users/timothee/git_clone/nim/timn/tests/nim/all/t11246.nim(21) main
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/uri.nim(193) fn1

```


=> otherwise you loose where the exception is raised from (/Users/timothee/git_clone/nim/Nim_prs/lib/pure/uri.nim(193) fn1)


note that this reveals a nim bug (unless I'm missing something) which we should file, which is that parent exceptions aren't shown by default; nim should use something like `exceptionToString` to show parents

Reply:
Since it is a bug, let's solve it later after this has been merged?

Reply:
there are other ways to show error messages, https://github.com/nim-lang/Nim/issues/15868 is just the default catch-all; including `parent` is harmless.

but if for some reason you don't want to do it in this PR, not a blocker